sycl/test-e2e/VirtualFunctions/misc/group-barrier.cpp

```diff
@@ -0,0 +1,140 @@
+      for (size_t I = 0; I < It.get_local_range().size(); ++I) {
+        Res += LocalData[I];
+      }
+    }
+
+    return Res;
+  }
+};
+
+class MultiplyOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::nd_item<1> It) {
+    LocalData[It.get_local_id()] += It.get_local_id();
+    sycl::group_barrier(It.get_group());
+    int Res = 1;
+    if (It.get_group().leader()) {
+      for (size_t I = 0; I < It.get_local_range().size(); ++I) {
+        Res *= LocalData[I];
+      }
+    }
+
+    return Res;
+  }
+};
+
+int main() try {
+  using storage_t = obj_storage_t<SumOp, MultiplyOp>;
+
+  auto asyncHandler = [](sycl::exception_list list) {
+    for (auto &e : list)
+      std::rethrow_exception(e);
+  };
+
+  sycl::queue q(asyncHandler);
+
+  auto *DeviceStorage = sycl::malloc_shared<storage_t>(1, q);
+  sycl::range G{512};
+  sycl::range L{32};
+
+  constexpr oneapi::properties props{oneapi::assume_indirect_calls};
+  for (unsigned TestCase = 0; TestCase < 2; ++TestCase) {
+    std::vector<int> HostData(G.size());
+    std::iota(HostData.begin(), HostData.end(), 0);
+    std::vector<int> DeviceData = HostData;
+    sycl::buffer<int> DataStorage(DeviceData.data(), G);
+
+    q.submit([&](sycl::handler &CGH) {
+       CGH.single_task([=]() {
+         DeviceStorage->construct</* ret type = */ BaseOp>(TestCase);
+       });
+     }).wait_and_throw();
+
+    q.submit([&](sycl::handler &CGH) {
+      sycl::accessor DataAcc(DataStorage, CGH, sycl::read_write);
+      sycl::local_accessor<int> LocalAcc(L, CGH);
+      CGH.parallel_for(sycl::nd_range{G, L}, props, [=](auto It) {
+        LocalAcc[It.get_local_id()] = DataAcc[It.get_global_id()];
+        auto *Ptr = DeviceStorage->getAs<BaseOp>();
+        DataAcc[It.get_global_id()] = Ptr->apply(
+            LocalAcc.template get_multi_ptr<sycl::access::decorated::no>()
+                .get(),
+            It);
+      });
+    });
+
+    // We can't call group_barrier on host and therefore here we have a
+    // reference function instead of calling the same methods on host.
+    for (size_t GID = 0; GID < G.size() / L.size(); ++GID) {
+      for (size_t LID = 0; LID < L.size(); ++LID)
+        HostData[GID * L.size() + LID] += LID;
+
+      int Res = (TestCase == 0) ? 0 : 1;
+      for (size_t LID = 0; LID < L.size(); ++LID) {
+        if (TestCase == 0)
+          Res += HostData[GID * L.size() + LID];
+        else
+          Res *= HostData[GID * L.size() + LID];
+      }
+
+      for (size_t LID = 0; LID < L.size(); ++LID)
+        HostData[GID * L.size() + LID] = Res;
+    }
+
+    sycl::host_accessor HostAcc(DataStorage);
+    for (size_t I = 0; I < HostData.size(); ++I)
+      assert(HostAcc[I] == HostData[I]);
+  }
+
+  sycl::free(DeviceStorage, q);
+
+  return 0;
+} catch (sycl::exception &e) {
+  std::cout << "Unexpected exception was thrown: " << e.what() << std::endl;
+  return 1;
+}
```

Comment:
This test generates two device images. One contains definitions of indirectly-callable functions and the other one contains kernel functions. In AOT mode they are not linked together before calling opencl-aot. When will this be fixed?

Reply:
Yeah, AOT support for virtual functions is incomplete yet, that will be addressed in separate PRs. For now that's a second priority, because there is plenty enough bugs even on JIT path